File: src/binary_heap.c
//
//  binary_heap.c
//  c-collections
//
//  Created by Fang Ling on 2023/2/14.
//

#include "binary_heap.h"

#define var __auto_type

/** Begin: Creating a heap **/
/* Creates a new, empty heap. */
struct BinaryHeap* binary_heap_init(Int element_size,
                                    Int (*compare)(const void* lhs,
                                                   const void* rhs)) {
    var heap = (struct BinaryHeap*)malloc(sizeof(struct BinaryHeap));
    heap -> count = 0;
    heap -> is_empty = true;
    heap -> compare = compare;
    heap -> storage = array_init(element_size);

    return heap;
}
/** End: Creaing a heap **/

/** Begin: Destroying a heap **/
/* Destroys all the elements and deallocates all the storage capacity. */
void binary_heap_deinit(struct BinaryHeap* heap) {
    array_deinit(heap -> storage);
    free(heap);
}
/** End: Destroying a heap **/

/** Begin: Private helpers **/
/* Index calculation functions. */
static inline Int left(Int i)   { return 2 * i + 1; }   /* Left child */
static inline Int right(Int i)  { return 2 * i + 2; }   /* Right child */
static inline Int parent(Int i) { return (i - 1) / 2; } /* Parent */

/* Maintain the max heap property from node i all the way up to root. */
static void max_heapify_up(struct BinaryHeap* heap, Int i) {
    if (i == 0) { /* we are at root node now */
        return;
    }
    /* if parent(i).key <= i.key */
    if (heap -> compare(array_get(heap -> storage, parent(i)),
                        array_get(heap -> storage, i)) <= 0) {
        /* Exchange i with parent(i) */
        array_swap_at(heap -> storage, i, parent(i));
        /* Current node is satisfy max heap property, but maybe not its
         * parent. Fix this recursively.
         */
        max_heapify_up(heap, parent(i));
    }
}

/* Maintain the max heap property from node i all the way down to leaf. */
static void max_heapify_down(struct BinaryHeap* heap, Int i) {
    var j = 0ll;
    /* Current i has both left and right children. */
    /* Let j be one of these children, we want to maximize j.key */
    if (left(i) < heap -> count && right(i) < heap -> count) {
        j = heap -> compare(array_get(heap -> storage, left(i)),
                            array_get(heap -> storage, right(i))) > 0
            ? left(i) : right(i);
    } else if (left(i) < heap -> count && right(i) >= heap -> count) {
        /* Node i only has left child. */
        j = left(i);
    } else if (left(i) >= heap -> count && right(i) < heap -> count) {
        /* Node i only has right child. */
        j = right(i);
    } else {
        /* Node i has zero child. i.e. Node i is a leaf node. We are done. */
        return;
    }
    /* If i is smaller than j (i's largest child), we are violate the max
     * heap property. Fix it by swap i.key with j.key and then recur on j.
     */
    if (heap -> compare(array_get(heap -> storage, i),
                        array_get(heap -> storage, j)) <= 0) {
        array_swap_at(heap -> storage, i, j);
        max_heapify_down(heap, j);
    }
}
/** End: Private helpers **/

/** Begin: Lookup **/
/* Returns the largest element in a heap in constant time. */
const void* binary_heap_max(struct BinaryHeap* heap) {
    if (heap -> is_empty) {
        return NULL;
    }
    return array_get(heap -> storage, 0);
}
/** End: Lookup **/

/** Begin: Insertion **/
/* Inserts a new item into heap. */
void binary_heap_insert(struct BinaryHeap* heap, void* key) {
    /* This works by adding the new element into the end of the backing array
     * and then bubbling it up to where it belongs in the heap.
     */
    heap -> is_empty = false;
    heap -> count += 1;
    array_append(heap -> storage, key);
    max_heapify_up(heap, heap -> count - 1);
}
/** End: Insertion **/

/** Begin: Removal **/
/* Removes the max item in the heap. */
void binary_heap_remove(struct BinaryHeap* heap) {
    /* To remove the largest element at index 0. Exchange it with the last item
     * in the array and then we trickle it down to where it belongs in the
     * heap.
     */
    if (heap -> count == 0) {
        fatal_error("Remove a item from a empty heap");
    }
    heap -> count -= 1;
    heap -> is_empty = heap -> count == 0 ? true : false;
    array_swap_at(heap -> storage, 0, heap -> count);
    max_heapify_down(heap, 0);
    array_remove_last(heap -> storage);
}
/** End: Removal **/

File: Sources/Collections/Heap/BinaryHeap.swift
//
//  BinaryHeap.swift
//
//
//  Created by Fang Ling on 2023/8/5.
//

import Foundation

//----------------------------------------------------------------------------//
//                         Structure Definition                               //
//----------------------------------------------------------------------------//
public struct BinaryHeap<Element : Comparable & Hashable> {
  @usableFromInline
  var _storage : [Element]
  
  /// A dictionary mapping elements to indices
  @usableFromInline
  var _dict : [Element : Int]
  
  //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++//
  //                     Inspection of a BinaryHeap                         //
  //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++//
  /// The number of elements in the binary heap.
  @inlinable
  public var count : Int { return _storage.count }
  
  /// A Boolean value indicating whether the binary heap is empty.
  @inlinable
  public var is_empty : Bool { return _storage.isEmpty }
  
  //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++//
  //                     Creation of a BinaryHeap                           //
  //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++//
  /// Creates a new, empty heap.
  @inlinable
  public init() {
    _storage = []
    _dict = [:]
  }
  
  /// Initializes a heap from a sequence.
  /// - Complexity: O(n), where n is the number of items in `elements`.
  @inlinable
  public init<S : Sequence>(_ elements : S) where S.Element == Element {
    _storage = Array(elements)
    _dict = [:]
    guard count > 1 else { return }
    
    for i in _storage.indices {
      _dict[_storage[i]] = i
    }
    for i in stride(from: count / 2 - 1, through: 0, by: -1) {
      max_heapify_down(i)
    }
  }
}

extension BinaryHeap: ExpressibleByArrayLiteral {
  /// Creates a new heap from the contents of an array literal.
  ///
  /// **Do not call this initializer directly.** It is used by the compiler when
  /// you use an array literal. Instead, create a new heap using an array
  /// literal as its value by enclosing a comma-separated list of values in
  /// square brackets. You can use an array literal anywhere a heap is expected
  /// by the type context.
  ///
  /// - Parameter elements: A variadic list of elements of the new heap.
  public init(arrayLiteral elements: Element...) {
    self.init(elements)
  }
}

//----------------------------------------------------------------------------//
//                         Index calculation                                  //
//----------------------------------------------------------------------------//
@inlinable
func bh_parent(_ i : Int) -> Int { return (i - 1) / 2 }

@inlinable
func bh_left(_ i : Int) -> Int { return 2 * i + 1 }

@inlinable
func bh_right(_ i : Int) -> Int { return 2 * i + 2 }

extension BinaryHeap {
  @inlinable
  mutating func swap_at(_ i : Int, _ j : Int) {
    _dict[_storage[i]] = j
    _dict[_storage[j]] = i
    _storage.swapAt(i, j)
  }
}

//----------------------------------------------------------------------------//
//                         Maintenance of the BinaryHeap property             //
//----------------------------------------------------------------------------//
extension BinaryHeap {
  ///Maintain the max heap property from node i all the way up to root.
  @inlinable
  mutating func max_heapify_up(_ i : Int) {
    if i == 0 { /* We are at root node now. */
      return
    }
    if _storage[bh_parent(i)] <= _storage[i] {
      swap_at(bh_parent(i), i) /* Exchange i with parent(i) */
      /*
       * Current node is satisfy max heap property, but maybe not its
       * parent. Fix this recursively.
       */
      max_heapify_up(bh_parent(i))
    }
  }
  
  ///Maintain the max heap property from node i all the way down to leaf.
  @inlinable
  mutating func max_heapify_down(_ i : Int) {
    if i >= count / 2 { /* No children, we are done */
      return
    }
    var j = 0
    /* Find i's largest children j. */
    if bh_left(i) < count && bh_right(i) < count { /* Both children exist */
      j = _storage[bh_left(i)] > _storage[bh_right(i)] ? bh_left(i) : bh_right(i)
    } else if bh_left(i) < count && bh_right(i) >= count { /* Left only */
      j = bh_left(i)
    } else if bh_left(i) >= count && bh_right(i) < count { /* Right only */
      j = bh_right(i)
    } else { /* Will never happen */
      print("error")
      return
    }
    /*
     * If i is smaller than j (i's largest child), we are violate the max
     * heap property. Fix it by swap i.key with j.key and then recur on j.
     */
    if _storage[i] < _storage[j] {
      swap_at(i, j)
      max_heapify_down(j)
    }
  }
}

//----------------------------------------------------------------------------//
//                         Lookup                                             //
//----------------------------------------------------------------------------//
extension BinaryHeap {
  /// Returns the largest element in a heap in constant time.
  @inlinable
  public var max : Element? {
    return _storage.first
  }
}

//----------------------------------------------------------------------------//
//                         Insertion                                          //
//----------------------------------------------------------------------------//
extension BinaryHeap {
  /// Inserts a new item into heap.
  /// - Complexity: O(lg n), where n is the length of the heap
  @inlinable
  public mutating func insert(_ new_element : Element) {
    _storage.append(new_element)
    _dict[new_element] = count - 1
    max_heapify_up(count - 1)
  }
}

//----------------------------------------------------------------------------//
//                         Removal                                            //
//----------------------------------------------------------------------------//
extension BinaryHeap {
  /// Removes the max item in the heap.
  @inlinable
  @discardableResult
  public mutating func remove_max() -> Element {
    swap_at(0, count - 1)
    /* It's necessary to first do remove to update `count` */
    let r = _storage.removeLast()
    _dict.removeValue(forKey: r)
    max_heapify_down(0)
    return r
  }
}

//----------------------------------------------------------------------------//
//                        Accessing Elements                                  //
//----------------------------------------------------------------------------//
extension BinaryHeap {
  /// - Complexity: O(lg n), where n is the length of the heap
  @inlinable
  public mutating func replace(_ old_element : Element, with element : Element) {
    guard let i = _dict[old_element] else {
      fatalError("no such element: \(old_element)")
    }
    if element < _storage[i] {
      fatalError("new element is smaller than current element")
    }
    
    _storage[i] = element
    max_heapify_up(i)
  }
}
